gtsam/sfm/PowerMethod.h

```diff
@@ -0,0 +1,367 @@
+   *
+   */
+
+  // Const reference to an externally-held matrix whose minimum-eigenvalue we
+  // want to compute
+  const Sparse &A_;
+
+  const Matrix &S_;
+
+  const int m_n_; // dimension of Matrix A
+  const int m_nev_; // number of eigenvalues required
+
+  // flag for running power method or accelerated power method. If false, the former, vice versa.
+  bool accelerated_;
+
+  // a Polyak momentum term
+  double beta_;
+
+  // const int m_ncv_; // dimention of orthonormal basis subspace
+  size_t m_niter_; // number of iterations
+
+private:
+  Vector ritz_val_; // all ritz eigenvalues
+  Matrix ritz_vectors_; // all ritz eigenvectors
+  Vector ritz_conv_; // store whether the ritz eigenpair converged
+  Vector sorted_ritz_val_; // sorted converged eigenvalue
+  Matrix sorted_ritz_vectors_; // sorted converged eigenvectors
+
+public:
+ // Constructor
+ explicit PowerFunctor(const Sparse& A, const Matrix& S, int nev, int ncv,
+                       bool accelerated = false, double beta = 0)
+     : A_(A),
+       S_(S),
+       m_n_(A.rows()),
+       m_nev_(nev),
+       // m_ncv_(ncv > m_n_ ? m_n_ : ncv),
+       accelerated_(accelerated),
+       beta_(beta),
+       m_niter_(0) {
+         // Do nothing
+       }
+
+ int rows() const { return A_.rows(); }
+ int cols() const { return A_.cols(); }
+
+ // Matrix-vector multiplication operation
+ Vector perform_op(const Vector& x1, const Vector& x0) const {
+   // Do the multiplication
+   Vector x2 = A_ * x1 - beta_ * x0;
+   x2.normalize();
+   return x2;
+  }
+
+  Vector perform_op(const Vector& x1, const Vector& x0, const double beta) const {
+    Vector x2 = A_ * x1 - beta * x0;
+    x2.normalize();
+    return x2;
+  }
+
+  Vector perform_op(const Vector& x1) const {
+    Vector x2 = A_ * x1;
+    x2.normalize();
+    return x2;
+  }
+
+  Vector perform_op(int i) const {
+    if (accelerated_) {
+      Vector x1 = ritz_vectors_.col(i-1);
+      Vector x2 = ritz_vectors_.col(i-2);
+      return perform_op(x1, x2);
+    } else
+      return perform_op(ritz_vectors_.col(i-1));
+  }
+
+  long next_long_rand(long seed) {
+    const unsigned int m_a = 16807;
+    const unsigned long m_max = 2147483647L;
+    unsigned long lo, hi;
+
+    lo = m_a * (long)(seed & 0xFFFF);
+    hi = m_a * (long)((unsigned long)seed >> 16);
+    lo += (hi & 0x7FFF) << 16;
+    if (lo > m_max) {
+      lo &= m_max;
+      ++lo;
+    }
+    lo += hi >> 15;
+    if (lo > m_max) {
+      lo &= m_max;
+      ++lo;
+    }
+    return (long)lo;
+  }
+
+  Vector random_vec(const int len) {
+    Vector res(len);
+    const unsigned long m_max = 2147483647L;
+    for (int i = 0; i < len; i++) {
+      long m_rand = next_long_rand(m_rand);
+      res[i] = double(m_rand) / double(m_max) - double(0.5);
+    }
+    res.normalize();
+    return res;
+  }
+
+  /// Tuning the momentum beta using the Best Heavy Ball algorithm in Ref(3)
+  void setBeta() {
```

Comment:
Again, I would suggest that you break this out as a standalone function (implemented in the same file as the power method) -- possibly as a helper function that is called from inside the main powerMethod() function, if beta actually _needs_ to be estimated.   

[Note that it's possible to run the accelerated power method using a _pre-specified_ value of beta (i.e., without having to estimate a good value empirically).  This is actually the case for our intended application (computing the minimum eigenvalue of the certificate matrix) ... ]